### ISSUE TITLE
JSON File Import of Training Data?

### ISSUE THREAD

**Issue by mabramsBlythedale** [BrainJS/brain.js#437]
![meow](https://2.bp.blogspot.com/-8Xlol4MCT4g/WWwiuiz72NI/AAAAAAAAOho/rMzWHGYxQoM63HkGttfTNrkMS8H1VOPxwCLcBGAs/s1600/2qqcdAY.gif)

## *What* is wrong?
I'm trying to load a json file to use for the training data.  That's not working and I think it is because the JSON format may not correspond with what the train() function expects?

## *Where* does it happen?
I'm currently working with the LSTM() 

## *How* do we replicate the issue?
This is the code I am using:

<!DOCTYPE html>

<html>
<head runat="server">
    <meta charset="utf-8" />
    <title>Neural Network - Brain.js - Web Method Test</title>
    <style>
        .example {
            padding: 100px;
            background: darkred;
            color: white;
            font-size: 36px;
        }
    </style>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
    <script src="//unpkg.com/brain.js"></script>
</head>
<body>
    <form id="form1" runat="server">
        <div id="example" class="example">Example Text</div>
    </form>
</body>
</html>

<script>

    var example = document.querySelector("#example");
    const txtBrain = new brain.recurrent.LSTM();
    var jsn = [];
    var TrainingData
    var output = 'test'
    var FLAG_Complete

    document.addEventListener('DOMContentLoaded', function () {
        example.innerHTML = 'Test';

        $.getJSON(['data.json'], function (jsn) {
            console.log(jsn.data);
            FLAG_Complete = true
        });

        setTimeout(function () {
            if (FLAG_Complete == true) {
                console.log(FLAG_Complete);
                runRequest();
            }
        }, 3000);

    }, false);

    function runRequest() {

        FLAG_Complete = false

        TrainingData = JSON.stringify(jsn);
        txtBrain.train(TrainingData, {
            log: details => console.log(details)
        });

        FLAG_Complete = true

        console.log('training complete');

        if (FLAG_Complete = true) {
            console.log('test drive brain');

            output = txtBrain.run('Test drive');
        }

        console.log('Category: ' + output);
        example.innerHTML = 'Category: ' + output;

    }

</script>

Here is the JSON file data:
[
  {
    "input": "test has failed",
    "output": "software"
  },
  {
    "input": "test has succeeded",
    "output": "software"
  },
  {
    "input": "I got an error",
    "output": "software"
  },
  {
    "input": "There is a bug",
    "output": "software"
  },
  {
    "input": "the power supply is faulty",
    "output": "hardware"
  },
  {
    "input": "I need more RAM",
    "output": "hardware"
  },
  {
    "input": "I ran out of disk space",
    "output": "hardware"
  },
  {
    "input": "unit testing is completed",
    "output": "software"
  },
  {
    "input": "the power supply is working correctly",
    "output": "hardware"
  },
  {
    "input": "memory is running low",
    "output": "hardware"
  },
  {
    "input": "code",
    "output": "software"
  },
  {
    "input": "bug",
    "output": "software"
  },
  {
    "input": "fault",
    "output": "hardware"
  },
  {
    "input": "crash",
    "output": "software"
  },
  {
    "input": "broke",
    "output": "hardware"
  }
]

The error I am seeing in the console.log for the training details is:

iterations: 18970, training error: Infinity

I am assuming that the JSON format of my file is incorrect.  Any thoughts on how I can do this would be much appreciated!  Thanks!!   Brain.js is awesome.

## *How* important is this (1-5)?
I'm just trying to learn so a 1 is fine.

## Expected behavior (i.e. solution)
I was hoping to see the train() process work without errors.

## Other Comments

**Comment by mabramsBlythedale** [comment]
PS ... I got started with this following this person's example here:  https://www.youtube.com/watch?v=RVMHhtTqUxc  ... since I ran into this error I tried changing the method of importing the JSON file, and also changed the element names to correspond with what the training data looks like it may want ... to no avail.

**Comment by mabramsBlythedale** [comment]
Ok... my fundamentals were not so good, and I was making several mistakes which didn't have to do with the train() function.  I got past those and am happily working on the next leg.  No need to respond to this.  For those who might be interested or find it useful the method that wound up working in a pure browser based javascript environment (not using node.js) is this...
```
<!DOCTYPE html>

<html>
<head runat="server">
    <meta charset="utf-8" />
    <title>Neural Network - Brain.js - Web Method Test</title>
    <style>
        div {
            padding: 12px;
            background-color: aliceblue;
        }

        .txt {
            width: 400px;
        }

        .info {
            font-size: 16px;
        }

        .btn {
            margin: 12px;
            height: 24px;
            width: 50px;
            padding: 6px;
        }

        .entry {
            padding: 30px;
            background: lightblue;
            color: black;
            font-size: 36px;
        }

        .example {
            padding: 100px;
            background: darkred;
            color: white;
            font-size: 36px;
        }

        .header {
            padding 22px;
            background: plum;
            font-size: 48px;
            color: black;
        }
    </style>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
    <script src="//unpkg.com/brain.js"></script>
</head>
<body>
    <form id="form1" runat="server">
        <div class="header">
            Brain.JS Example
        </div>
        <div class="entry">
            <span class="info">How many Iterations to train (poor but fast: 100, good but slow: 4000)?</span>
            <input type="text" id="txtIterations" value="1000" />
            <br />
            <input type="button" value="Train the AI" onclick="javascript: TrainBrain();" />
        </div>
        <div>
            <span class="info">What is your computer issue?</span>
            <br />
            <input type="text" id="txtInput" class="txt" />
            <br />
            <span class="info">(will answer 'softare' or 'hardware') </span>
            <br />
            <br />
            <input type="button" value="Send Test" onclick="javascript: runRequest();" />
            <br />
        </div>
        <div id="example" class="example">Example Text</div>
    </form>
</body>
</html>

<script>
    const txtBrain = new brain.recurrent.LSTM();
    var example = document.querySelector("#example");

    document.addEventListener('DOMContentLoaded', function () {
        example.innerHTML = 'The Brain is Waiting to be Trained...';
        console.log('here');
    }, false);

    function TrainBrain() {
        example.innerHTML = "Training in progress... please wait"
        console.log('inside TrainBrain');
        var jsn = [];
        var TrainingData
        var TData = [];
        var txtIterations = document.getElementById("txtIterations").value

        setTimeout(function () {

            console.log('inside timeout');

            $.getJSON(['data.json'], function (jsn) {

                TrainingData = JSON.stringify(jsn);

                TData = JSON.parse(TrainingData);


                var TData2 = TData.map(item => ({
                    input: item.input,
                    output: item.output
                }))
                const iterations = txtIterations
                const config = {
                    iterations: iterations
                    , log: details => console.log(details)
                };
                txtBrain.train(TData2, config);
                example.innerHTML = 'Training is completed. Please ask your question.'
            });
        }, 2000);

        console.log('after timeout');
    }

    function runRequest() {
        var output = ''
        var txtInput = document.getElementById("txtInput").value

        console.log(txtInput);
        console.log('inside run request')

        output = txtBrain.run(txtInput.text);
        console.log('Category: ' + output);
        example.innerHTML = txtInput + '<br>... Category: ' + output;

    }

</script>
```

**Comment by robertleeplummerjr** [comment]
Looking at this now.

**Comment by robertleeplummerjr** [comment]
Reproduced a working version here: https://jsfiddle.net/robertleeplummerjr/w1f4nmyz/6/
There was an issue in alpha 4 (I think) that was making learning terribly impossible, which has been fixed.  Now the net learns seemingly immediate, my CPU doesn't even rise.  I believe that to be the issue here because of this: `iterations: 18970, training error: Infinity`, which is essentially "[vanishing and exploding gradient](https://en.wikipedia.org/wiki/Vanishing_gradient_problem)".  Any time your network gives you that high of a training error, either you are using a vanilla RNN, it is a configuration issue, or the net is simply not implemented correctly.

Anyway, ty for experimenting and seeing it through, and please let me know if a later verison (alpha 6, just released a few minutes ago) fixes your problem.  Closing as I believe we have identified the issue, and a solid fix.
